rust/src/push/evaluator.rs

```diff
@@ -0,0 +1,330 @@
+
+                let sender_pattern = if let Some(sender) = sender {
+                    Some(sender.as_ref())
+                } else if let Some(sender_type) = sender_type {
+                    if sender_type == "user_id" {
+                        if let Some(user_id) = user_id {
+                            Some(user_id)
+                        } else {
+                            return Ok(false);
+                        }
+                    } else {
+                        warn!("Unrecognized sender_type: {sender_type}");
+                        return Ok(false);
+                    }
+                } else {
+                    None
+                };
+
+                let sender_compiled_pattern = if let Some(pattern) = sender_pattern {
+                    Some(glob_to_regex(pattern, GlobMatchType::Whole)?)
+                } else {
+                    None
+                };
+
+                let type_compiled_pattern = if let Some(pattern) = event_type_pattern {
+                    Some(glob_to_regex(pattern, GlobMatchType::Whole)?)
+                } else {
+                    None
+                };
+
+                for (relation_sender, event_type) in relations {
+                    if let Some(pattern) = &sender_compiled_pattern {
+                        if !pattern.is_match(relation_sender) {
+                            continue;
+                        }
+                    }
+
+                    if let Some(pattern) = &type_compiled_pattern {
+                        if !pattern.is_match(event_type) {
+                            continue;
+                        }
+                    }
+
+                    return Ok(true);
+                }
+
+                false
+            }
+        };
+
+        Ok(result)
+    }
+
+    fn match_event_match(
+        &self,
+        event_match: &EventMatchCondition,
+        user_id: Option<&str>,
+    ) -> Result<bool, Error> {
+        let pattern = if let Some(pattern) = &event_match.pattern {
+            pattern
+        } else if let Some(pattern_type) = &event_match.pattern_type {
+            let user_id = if let Some(user_id) = user_id {
+                user_id
+            } else {
+                return Ok(false);
+            };
+            match &**pattern_type {
+                "user_id" => user_id,
+                "user_localpart" => get_localpart_from_id(user_id)?,
+                _ => return Ok(false),
+            }
+        } else {
+            return Ok(false);
+        };
+
+        // For the content.body we match against "words", but for everything
+        // else we match against the entire value.
+        let match_type = if event_match.key == "content.body" {
+            GlobMatchType::Word
+        } else {
+            GlobMatchType::Whole
+        };
+
+        if let Some(value) = self.flattened_keys.get(&*event_match.key) {
+            let compiled_pattern = glob_to_regex(pattern, match_type)?;
```

Comment:
Is a cache of compiled patterns would be useful ? Compiling regex tends to be expensive.

Reply:
Oh, good point. We did have a cache before too. Let me quickly run a benchmark.

Reply:
Interesting numbers come out of it (annoying I can't commit the benchmark code, as benchmarking doesn't work in a `cdylib` crate):

```
test bench_glob_whole   ... bench:      28,814 ns/iter (+/- 1,209)
test bench_glob_word    ... bench:     409,264 ns/iter (+/- 47,114)
```

So compiling word based regex can take almost half a millisecond on my laptop. If we disable unicode awareness when compiling the regex they end up taking about the same amount of time.

Reply:
Instead of caching the regex, I've implemented a "fast path", which allows us to avoid compiling the regex unless necessary.